admin/tests/unit/controllers/targets-edit.js

```diff
@@ -0,0 +1,86 @@
+describe('TargetsEditCtrl controller', () => {
+  'use strict';
+
+  let dbGet,
+    mockTargetsUnique,
+    scope,
+    Settings,
+    translate,
+    UpdateSettings;
+
+  beforeEach(() => {
+    module('adminApp');
+
+    dbGet = sinon.stub().returns(
+      Promise.resolve(
+        ['medic-clinic','medic-chw']
+      )
+    );
+    Settings = sinon.stub().returns(
+      Promise.resolve({
+        locales: [
+          {code: 'en', name: 'English'}
+        ],
+        tasks: {
+          targets: {
+            items: [
+              {type: 'count', id: 'TaskId', icon: 'medic-clinic', goal: 'goal', name: { content: 'content' }},
+              {type: 'count', id: 'TaskId2', icon: 'medic-chw', goal: 'goal', name: { content: 'content' }}
+            ]
+          }
+        }
+      })
+    );
+    translate = sinon.stub();
+    UpdateSettings = sinon.stub();
+
+    module($provide => {
+      $provide.factory(
+        'DB',
+        KarmaUtils.mockDB({
+          get: dbGet,
+        })
+      );
+      $provide.value('Settings', Settings);
+      $provide.value('translate', translate);
+      $provide.value('UpdateSettings', UpdateSettings);
+    });
+
+    inject((translate, $controller) => {
+      const createController = () => {
+        scope = {};
+        return $controller('TargetsEditCtrl', {
+          $q: Q,
+          $scope: scope,
+          $translate: translate
+        }); 
+      };
+      mockTargetsUnique = () => {
+        createController();
+      };
+    });
+  });
+
+  afterEach(() => {
+    KarmaUtils.restore(
+      dbGet,
+      Settings,
+      UpdateSettings
+    );
+  });
+
+  describe('targets edit', () => {
+    it ('validation failure, target id not unique', done => {
+      mockTargetsUnique();
+      setTimeout(() => {
+        scope.target = {type: 'count', id: 'TaskId', icon: 'medic-clinic', goal: 'goal'};
+        scope.submit();
+        setTimeout(() => {
```

Comment:
Ideally you wouldn't need `setTimeout` - these may not be needed any more once the async functions above are made to be sync.

Reply:
Well these two `setTimeout` are calls to `DB()` and `Settings()` which are async in nature

Reply:
Right, but you've mocked them with sinon so in the scope of this test they are synchronous. Our other tests don't use timeouts so you should be able to get rid of them here too.

Reply:
Well the sinon  stub as resolving with a Promise, if I remove the `setTimeout` the tests fail